File: P3.py
# -*- coding: utf-8 -*-
"""
Created on Mon Apr  5 16:02:29 2021

@author: jgdch
"""

#Space complexity: O(N), where N is the number of elements in seen
#Time Complexity: O(N), where N is the length of the input matrix
class Solution(object):
    def spiralOrder(self, matrix):
        if not matrix:
            return []
        m, n = len(matrix), len(matrix[0])
        seen = [[0] * n for i in matrix]
        out= []
        dr = [0, 1, 0, -1]
        dc = [1, 0, -1, 0]
        r = c = 0
        direc = 0
        for i in range(m * n):
            
            out.append(matrix[r][c])
            seen[r][c] = 1
            mr, mc = r + dr[direc], c + dc[direc]
            if 0 <= mr < m and 0 <= mc < n and not seen[mr][mc]:
                r, c = mr, mc
            else:
                direc = (direc + 1) % 4
                r, c = r + dr[direc], c + dc[direc]
        return out

File: p3.java
// Time Complexity : O(m*n)
// Space Complexity : O(1)
// Did this code successfully run on Leetcode : Yes
// Three line explanation of solution in plain english
// Traversing through the matrix in the order shown in the question by considering on spiral at a time. Maintaining four pointers ie left, right, up and down to define the boundaries of the spiral. Adding elements into the list while traversing through the spiral defined by boundaries

// Your code here along with comments explaining your approach

class Solution {
    public List<Integer> spiralOrder(int[][] matrix) {
        // Output list
        List<Integer> result = new ArrayList<>();
        // Left boundary
        int l = 0;
        // Right boundary
        int r = matrix[0].length - 1;
        // Upper boundary
        int u = 0;
        // Lower boundary
        int d = matrix.length - 1;

        while (d >= u && r >= l) {

            // Traverisng through the spiral
            // Going from left to right
            for (int i = l; i <= r; i++) {
                result.add(matrix[u][i]);
            }
            u++;
            // Going from up to down
            for (int i = u; i <= d; i++) {
                result.add(matrix[i][r]);
            }
            r--;
            // Going from right to left
            if (d >= u) {
                for (int i = r; i >= l; i--) {
                    result.add(matrix[d][i]);
                }
                d--;
            }
            // Going from down to up
            if (r >= l) {
                for (int i = d; i >= u; i--) {
                    result.add(matrix[i][l]);
                }
                l++;
            }
        }

        return result;
    }
}